### PR TITLE
Add MPI streams

### PR DESCRIPTION
**Pull request by biddisco**
This work is not completely finished - I can still get deadlocks with the async storage test, but I would like to start the process of merging this, so I can test it more and use dlaf as a more advanced testing framework

```
Support throttling of mpi traffic using streams
    
Throttling using a single counter causes deadlocks as a
receive might be blocked by a send, or vice versa.
 ```
+ Some flyby cmake and cleanup fixes

### CODE REVIEW

File: libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp [hunk 1]
```diff
@@ -26,17 +26,29 @@
...
 namespace pika::mpi::experimental {
+
+    enum stream_index : std::uint32_t
+    {
+        default_stream = 0,
+        send_stream,
+        receive_stream,
+        collective_stream,
+        user_stream,
+    };
```
Comment (msimberg) [inline libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp:37]: Maybe something like this?
```suggestion
    enum class stream_type : std::uint32_t // does the user need to know that it's used as an index? it's the type of stream that they're interested in, right?
    {
        default_ = 0, // possibly different name here to avoid the underscore...
        send,
        receive,
        collective,
        user,
    };
```

File: libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp [hunk 2]
```diff
@@ -26,17 +26,29 @@
...
+
     namespace detail {
         // -----------------------------------------------------------------
         // by convention the title is 7 chars (for alignment)
         using print_on = pika::debug::detail::enable_print<false>;
         static constexpr print_on mpi_debug("MPI_FUT");
+        struct mpi_queue;
```
Comment (msimberg) [inline libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp:44]: Is it a stream or a queue? I don't mind which one, but it should be used consistently.

File: libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp [hunk 3]
```diff
@@ -26,17 +26,29 @@
...
 
         using request_callback_function_type =
             pika::util::detail::unique_function<void(int)>;
 
         PIKA_EXPORT void add_request_callback(
-            request_callback_function_type&& f, MPI_Request req);
+            request_callback_function_type&& f, MPI_Request req,
+            stream_index queue);
```
Comment (msimberg) [inline libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp:51]: Same here.

File: libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp [hunk 4]
```diff
@@ -61,17 +74,18 @@ namespace pika::mpi::experimental {
         /// This should be used with great caution as setting it too low can
         /// cause deadlocks. The default value is size_t(-1) - i.e. unlimited
         /// The value can be set using an environment variable as follows
-        /// PIKA_MPI_MSG_THROTTLE=512
+        /// PIKA_MPI_MSG_THROTTLE=64
         /// but user code setting it will override any default or env value
         /// This function returns the previous throttling threshold value
-        PIKA_EXPORT size_t set_max_requests_in_flight(size_t);
+        PIKA_EXPORT std::uint32_t set_max_requests_in_flight(
+            std::uint32_t, stream_index = stream_index(-1));
```
Comment (msimberg) [inline libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp:81]: What does the `-1` represent? Could we use an optional instead?

File: libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp [hunk 5]
```diff
@@ -61,17 +74,18 @@ namespace pika::mpi::experimental {
...
 
         /// Query the current value of the throttling threshold
-        PIKA_EXPORT size_t get_max_requests_in_flight();
+        PIKA_EXPORT std::uint32_t get_max_requests_in_flight(stream_index);
 
         // -----------------------------------------------------------------
         /// returns the number of mpi requests currently outstanding
-        PIKA_EXPORT size_t get_num_requests_in_flight();
+        PIKA_EXPORT std::uint32_t get_num_requests_in_flight();
```
Comment (msimberg) [inline libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp:88]: Why `uint32_t` instead of `size_t`? Not that we'll probably ever have more than `2^32` requests in flight at the same time... but still.

File: libs/pika/async_mpi/include/pika/async_mpi/transform_mpi.hpp [hunk 6]
```diff
@@ -383,22 +389,49 @@ namespace pika::mpi::experimental {
...
+        template <typename Sender, typename F,
+            PIKA_CONCEPT_REQUIRES_(
+                pika::execution::experimental::is_sender_v<Sender>)>
+        friend constexpr PIKA_FORCEINLINE auto tag_fallback_invoke(
+            transform_mpi_t, Sender&& sender, F&& f,
+            mpi::experimental::stream_index s)
+        {
+            return transform_mpi_detail::transform_mpi_sender<Sender, F>{
+                PIKA_FORWARD(Sender, sender), PIKA_FORWARD(F, f), s};
+        }
```
Comment (msimberg) [inline libs/pika/async_mpi/include/pika/async_mpi/transform_mpi.hpp:414]: Isn't one overload enough?
```suggestion
        template <typename Sender, typename F,
            PIKA_CONCEPT_REQUIRES_(
                pika::execution::experimental::is_sender_v<Sender>)>
        friend constexpr PIKA_FORCEINLINE auto tag_fallback_invoke(
            transform_mpi_t, Sender&& sender, F&& f,
            stream_index s = stream_index::default_)
        {
            return transform_mpi_detail::transform_mpi_sender<Sender, F>{
                PIKA_FORWARD(Sender, sender), PIKA_FORWARD(F, f), s};
        }
```

File: libs/pika/async_mpi/include/pika/async_mpi/transform_mpi.hpp [hunk 7]
```diff
@@ -383,22 +389,49 @@ namespace pika::mpi::experimental {
...
         }
+
+        template <typename F>
+        friend constexpr PIKA_FORCEINLINE auto tag_fallback_invoke(
+            transform_mpi_t, F&& f, mpi::experimental::stream_index s)
+        {
+            return ::pika::execution::experimental::detail::partial_algorithm<
+                transform_mpi_t, F, mpi::experimental::stream_index>{
+                PIKA_FORWARD(F, f), s};
+        }
```
Comment (msimberg) [inline libs/pika/async_mpi/include/pika/async_mpi/transform_mpi.hpp:434]: Same here, isn't one overload enough?

File: libs/pika/async_mpi/src/mpi_polling.cpp [hunk 8]
```diff
@@ -26,14 +26,18 @@
 #include <vector>
 
 namespace pika::mpi::experimental {
+
+    constexpr std::uint32_t max_mpi_streams = 4;
```
Comment (msimberg) [inline libs/pika/async_mpi/src/mpi_polling.cpp:30]: Could this instead be the last member of the `stream_index` enum so that you don't have to keep this in sync with the enum?

File: libs/pika/async_mpi/src/mpi_polling.cpp [hunk 9]
```diff
@@ -26,14 +26,18 @@
...
         // -----------------------------------------------------------------
         /// Holds an MPI_Request and a callback. The callback is intended to be
         /// called when the operation tied to the request handle completes.
         struct request_callback
         {
-            MPI_Request request;
-            request_callback_function_type callback_function;
+            MPI_Request request_;
+            request_callback_function_type callback_function_;
+            stream_index index_{0};
```
Comment (msimberg) [inline libs/pika/async_mpi/src/mpi_polling.cpp:40]:
```suggestion
            stream_index index_{stream_index::default};
```
is maybe more obvious?

File: libs/pika/async_mpi/src/mpi_polling.cpp [hunk 10]
```diff
@@ -53,7 +57,23 @@ namespace pika::mpi::experimental {
...
+        struct mpi_stream
+        {
+            mutex_type throttling_mtx_;
+            pika::lcos::local::condition_variable throttling_cond_;
+            std::atomic<std::uint32_t> in_flight_{0};
+            std::uint32_t limit_{get_throttling_default()};
+            std::uint32_t index;
+        };
+
+        using mpi_cbq_tuple =
```
Comment (msimberg) [inline libs/pika/async_mpi/src/mpi_polling.cpp:75]:
```suggestion
        using mpi_callback_queue_tuple =
```
?

File: libs/pika/async_mpi/src/mpi_polling.cpp [hunk 11]
```diff
@@ -221,17 +263,20 @@ namespace pika::mpi::experimental {
 
             // Eagerly check if request already completed. If it did, call the
             // callback immediately.
+#ifdef EAGER_CHECK
```
Comment (msimberg) [inline libs/pika/async_mpi/src/mpi_polling.cpp:266]:
```suggestion
#ifdef PIKA_MPI_EAGER_CHECK
```
? Or if this is just for debugging please remove it.

File: libs/pika/async_mpi/tests/unit/mpi_async_storage.cpp [hunk 12]
```diff
@@ -235,7 +235,9 @@ void test_send_recv(std::uint32_t rank, std::uint32_t nranks, std::mt19937& gen,
             void* buffer_to_recv = &local_recv_storage[memory_offset_recv];
             auto rsnd = ex::just(buffer_to_recv, options.transfer_size_B,
                             MPI_UNSIGNED_CHAR, recv_rank, tag, MPI_COMM_WORLD) |
-                mpi::transform_mpi(MPI_Irecv) | ex::then([&](int result) {
+                mpi::transform_mpi(
+                    MPI_Irecv, /*mpi::mpi_stream::*/ mpi::receive_stream) |
```
Comment (msimberg) [inline libs/pika/async_mpi/tests/unit/mpi_async_storage.cpp:239]: Is the commented-part still needed?

File: libs/pika/async_mpi/tests/unit/mpi_async_storage.cpp [hunk 13]
```diff
@@ -258,7 +260,9 @@ void test_send_recv(std::uint32_t rank, std::uint32_t nranks, std::mt19937& gen,
             void* buffer_to_send = &local_send_storage[memory_offset_send];
             auto ssnd = ex::just(buffer_to_send, options.transfer_size_B,
                             MPI_UNSIGNED_CHAR, send_rank, tag, MPI_COMM_WORLD) |
-                mpi::transform_mpi(MPI_Isend) | ex::then([&](int result) {
+                mpi::transform_mpi(
+                    MPI_Isend, mpi::/*mpi_stream::*/ send_stream) |
```
Comment (msimberg) [inline libs/pika/async_mpi/tests/unit/mpi_async_storage.cpp:264]: Same here?

File: libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp [hunk 14]
```diff
@@ -82,11 +82,13 @@ int pika_main(pika::program_options::variables_map& vm)
 
     // if comm size < 2 this test should fail
     // it needs to run on N>2 ranks to be useful
-    PIKA_TEST_MSG(size > 1, "This test requires N>1 mpi ranks");
+    //PIKA_TEST_MSG(size > 1, "This test requires N>1 mpi ranks");
```
Comment (msimberg) [inline libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp:85]: Should this still be commented out?

File: libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp [hunk 15]
```diff
@@ -188,12 +184,12 @@ int pika_main(pika::program_options::variables_map& vm)
...
-        {
-            std::cout << "Rank " << rank << " flight " << tt << " counter "
-                      << counter << std::endl;
-        }
+        //        auto tt = mpi::detail::get_num_requests_in_flight();
+        //        if (tt != 0)
+        //        {
+        //            std::cout << "Rank " << rank << " flight " << tt << " counter "
+        //                      << counter << std::endl;
+        //        }
```
Comment (msimberg) [inline libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp:192]: Debugging?

File: libs/pika/resource_partitioner/examples/oversubscribing_resource_partitioner.cpp [hunk 16]
```diff
@@ -86,18 +86,18 @@ int pika_main(/*pika::program_options::variables_map& vm*/)
         pika::execution::thread_priority::critical);
     pika::execution::parallel_executor normal_priority_executor;
 
-    pika::execution::parallel_executor mpi_executor;
+    pika::execution::parallel_executor pool_executor;
```
Comment (msimberg) [inline libs/pika/resource_partitioner/examples/oversubscribing_resource_partitioner.cpp:89]: These changes probably don't belong in this PR?

File: libs/pika/resource_partitioner/examples/simple_resource_partitioner.cpp [hunk 17]
```diff
@@ -71,19 +71,19 @@ int pika_main(pika::program_options::variables_map&)
         pika::execution::thread_priority::critical);
     pika::execution::parallel_executor normal_priority_executor;
 
-    pika::execution::parallel_executor mpi_executor;
+    pika::execution::parallel_executor pool_executor;
```
Comment (msimberg) [inline libs/pika/resource_partitioner/examples/simple_resource_partitioner.cpp:74]: These changes also probably don't belong here?

File: cmake/pika_add_test.cmake [hunk 18]
```diff
@@ -64,15 +64,17 @@ function(pika_add_test category name)
...
 
-  set(args "${${name}_UNPARSED_ARGUMENTS}" ${args})
+  set(args "${${name}_ARGS}" "${${name}_UNPARSED_ARGUMENTS}" ${args})
 
   set(_script_location ${PROJECT_BINARY_DIR})
 
   set(cmd ${_exe})
 
   if(${name}_RUNWRAPPER)
+    set(_preflags_list_ ${MPIEXEC_PREFLAGS})
```
Comment (msimberg) [inline cmake/pika_add_test.cmake:74]: Is `MPIEXEC_PREFLAGS` a `find_package(mpi)` thing or should we rather name it `PIKA_TEST_MPIEXEC_PREFLAGS` or something like that?

**Review by msimberg — CHANGES_REQUESTED** [review]
Looks good in general, and we don't have to be too careful with this since DLA-Future isn't using it yet. Mostly style comments.

File: libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp (hunk 1 above)
Comment (biddisco) [reply libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp:37]: I'm happy with that change - I used "index" in the name because if the user decided they needed more streams, then
`user_stream + 1`, `user_stream + 2`, etc might be needed. But I will worry about that some other time and see iif this is a viable way of doing things first.

File: libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp (hunk 2 above)
Comment (biddisco) [reply libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp:44]: it's a queue of stuff that we call a stream! I'll rename the vars appropriately.

File: libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp (hunk 4 above)
Comment (biddisco) [reply libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp:81]: the `-1` is the set `all` queues/streams with the same value. An optional is a good idea. I'll do that. (and add a comment to document why)

File: libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp (hunk 5 above)
Comment (biddisco) [reply libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp:88]: because in your changes, you changed one of my `size_t` into a `uint32_t` so I thought for some reason you preferred it. Perhaps that was a victim of my conflict resolution. I originally had `size_t` for them all. (though frankly 64 bits is probably overkill for these vars)

File: libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp (hunk 14 above)
Comment (biddisco) [reply libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp:85]: grrrr. I keep forgetting to undo those comments when I test locally with one rank. will fix

File: libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp (hunk 15 above)
Comment (biddisco) [reply libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp:192]: yup. forgot about that. Frankly I don't mind commented out debug stuff in tests as it helps people later when they try to finx some new bug they discover

File: libs/pika/resource_partitioner/examples/oversubscribing_resource_partitioner.cpp (hunk 16 above)
Comment (biddisco) [reply libs/pika/resource_partitioner/examples/oversubscribing_resource_partitioner.cpp:89]: no, but they are flyby fixes with "mpi" in them so I thought why-not.

File: cmake/pika_add_test.cmake (hunk 18 above)
Comment (biddisco) [reply cmake/pika_add_test.cmake:74]: its built into `find_package(mpi)`

File: libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp (hunk 5 above)
Comment (msimberg) [reply libs/pika/async_mpi/include/pika/async_mpi/mpi_polling.hpp:88]: If that's the case, my bad! I don't have any memory of changing them to `uint32_t`s... Do you have a link to what I changed? In any case, `uint32_t` or `size_t` are both fine by me.

File: libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp (hunk 15 above)
Comment (msimberg) [reply libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp:192]: Guard it with this instead of commenting it out: https://github.com/pika-org/pika/blob/cd738d191967f7f8656954372b93d8cc6e7997fd/libs/pika/async_mpi/tests/unit/mpi_ring_async_sender_receiver.cpp#L235?

File: libs/pika/resource_partitioner/examples/oversubscribing_resource_partitioner.cpp (hunk 16 above)
Comment (msimberg) [reply libs/pika/resource_partitioner/examples/oversubscribing_resource_partitioner.cpp:89]: Ah, right. I missed that it indeed doesn't make sense to call it `mpi_executor`. Fine, let's keep it here (same for the other one).

**Comment by biddisco** [comment]
Latest patch should address all comments. Moved a couple of functions to/from detail namespace since they might be considered part of public API. Will review these functions in an ongoing basis and possibly hide them as the module matures (throttling is not a finished product yet)

File: libs/pika/async_mpi/src/mpi_polling.cpp [hunk 19]
```diff
@@ -469,6 +514,48 @@ namespace pika::mpi::experimental {
         }
     }    // namespace detail
 
+    std::uint32_t set_max_requests_in_flight(
+        std::uint32_t N, std::optional<stream_type> s)
+    {
+        if (s == std::nullopt)
```
Comment (msimberg) [inline libs/pika/async_mpi/src/mpi_polling.cpp:520]: Optional (heh...):
```suggestion
        if (!s)
```
or
```suggestion
        if (!s.has_value())
```
I'm fine with the comparison to `nullopt` as well.

File: libs/pika/async_mpi/src/mpi_polling.cpp [hunk 20]
```diff
@@ -53,7 +61,23 @@ namespace pika::mpi::experimental {
...
+        {
+            mutex_type throttling_mtx_;
+            pika::lcos::local::condition_variable throttling_cond_;
+            std::atomic<std::uint32_t> in_flight_{0};
+            std::uint32_t limit_{get_throttling_default()};
+            std::uint32_t index;
+        };
+
+        using mpi_callback_queue_tuple =
+            std::tuple<request_callback_function_type, mpi_stream*>;
```
Comment (msimberg) [inline libs/pika/async_mpi/src/mpi_polling.cpp:80]: Seeing this again this would probably be better off in a struct, but that can be separately.

**Comment by msimberg** [comment]
bors try

**Review by msimberg — APPROVED** [review]
Thanks @biddisco for the updates. Two more optional comments that you're free to ignore.

File: libs/pika/async_mpi/include/pika/async_mpi/transform_mpi.hpp [hunk 21]
```diff
@@ -12,6 +12,7 @@
 #include <pika/config.hpp>
 #include <pika/assert.hpp>
 #include <pika/async_mpi/mpi_polling.hpp>
+#include <pika/async_mpi/mpi_throttling.hpp>
```
Comment (msimberg) [inline libs/pika/async_mpi/include/pika/async_mpi/transform_mpi.hpp:15]: This header doesn't exist by the way... All functionality is in `mpi_polling.hpp`, right?

File: libs/pika/async_mpi/CMakeLists.txt [hunk 22]
```diff
@@ -14,7 +14,7 @@ endif()
 # Default location is $PIKA_ROOT/libs/mpi/include
 set(async_mpi_headers
     pika/async_mpi/mpi_exception.hpp pika/async_mpi/mpi_polling.hpp
-    pika/async_mpi/transform_mpi.hpp
+    pika/async_mpi/mpi_throttling.hpp pika/async_mpi/transform_mpi.hpp
```
Comment (msimberg) [inline libs/pika/async_mpi/CMakeLists.txt:17]: Header doesn't exist.

**Review by msimberg — CHANGES_REQUESTED** [review]

File: libs/pika/async_mpi/CMakeLists.txt (hunk 22 above)
Comment (biddisco) [reply libs/pika/async_mpi/CMakeLists.txt:17]: oops. I started creating a new file with the throttling code in, but removed it's contents during my conflict merge and forgot to remove the header includes. Will fix.

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors merge

**Review by msimberg — APPROVED** [review]

**Comment by msimberg** [comment]
bors merge
